### PR TITLE
fix: add timeout handling to legacy cli

### PR DESCRIPTION
**Pull request by bastiandoetsch**
## Pull Request Submission

Please check the boxes once done.

The pull request must:

- **Reviewer Documentation**
    - [x] follow [CONTRIBUTING](https://github.com/snyk/cli/blob/master/CONTRIBUTING.md) rules
    - [x] be accompanied by a detailed description of the changes
    - [x] contain a risk assessment of the change (Low | Medium | High) with regards to breaking existing functionality. A change e.g. of an underlying language plugin can completely break the functionality for that language, but appearing as only a version change in the dependencies.
    - [ ] highlight breaking API if applicable
    - [x] contain a link to the automatic tests that cover the updated functionality.
    - [x] contain testing instructions in case that the reviewer wants to manual verify as well, to add to the manual testing done by the author.
    - [ ] link to the to the PR for the User-facing documentation
- **User facing Documentation**
    - [ ] update any relevant documentation in gitbook by submitting a gitbook PR, and including the PR link here
    - [ ] ensure that the message of the final single commit is descriptive and prefixed with either `feat:` or `fix:` , others might be used in rare occasions as well, if there is no need to document the changes in the release notes. The changes or fixes should be described in detail in the commit message for the changelog & release notes.
- **Testing**
    - [ ] Changes, removals and additions to functionality must be covered by acceptance / integration tests or smoke tests - either already existing ones, or new ones, created by the author of the PR.

## Pull Request Review

All pull requests must undergo a thorough review process before being merged. 
The review process of the code PR should include code review, testing, and any necessary feedback or revisions. 
Pull request reviews of functionality developed in other teams only review the given documentation and test reports. 

Manual testing will not be performed by the reviewing team, and is the responsibility of the author of the PR.

For Node projects: It’s important to make sure changes in `package.json` are also affecting `package-lock.json` correctly.

****************************If a dependency is not necessary, don’t add it.**************************** 

When adding a new package as a dependency, make sure that the change is absolutely necessary. We would like to refrain from adding new dependencies when possible.
Documentation PRs in gitbook are reviewed by Snyk's content team. They will also advise on the best phrasing and structuring if needed.

## Pull Request Approval

Once a pull request has been reviewed and all necessary revisions have been made, it is approved for merging into 
the main codebase. The merging of the code PR is performed by the code owners, the merging of the documentation PR 
by our content writers.

## What does this PR do?
This PR starts commands in the TypeScript CLI with a deadline, so that timeouts are actually killing spawned sub processes.

Also, the global timeout now waits for a grace-period of 5s on top of the timeout to exit. 

https://github.com/snyk/cli/assets/20150761/c16417cf-92ad-46ab-80c3-f23c99b21839

```
macbook-basti-lan :: ~/workspace/cli ‹fix/timeout-handling*› » SNYK_CI=1 SNYK_TIMEOUT_SECS=1 binary-releases/snyk-macos test --all-projects -d                              1 ↵
2023-11-28T11:14:31Z main - Version:               1.1256.0-dev.9067b9ee7f7407151c23f2bde2742ecc1f0c4440
2023-11-28T11:14:31Z main - Platform:              darwin amd64
2023-11-28T11:14:31Z main - API:                   https://api.snyk.io
2023-11-28T11:14:31Z main - Cache:                 /Users/bdoetsch/Library/Caches/snyk/snyk-cli
2023-11-28T11:14:31Z main - Organization:          7d7ab703-3630-4b46-afed-bad53a655ca3
2023-11-28T11:14:31Z main - Insecure HTTPS:        false
2023-11-28T11:14:31Z main - Analytics:             enabled
2023-11-28T11:14:31Z main - Authorization:         6f17318a293de207b1aea1f2052153d1[...]  (type=token)
2023-11-28T11:14:31Z main - Features:
2023-11-28T11:14:31Z main -   oauth:               Disabled
2023-11-28T11:14:31Z main -   fips:                Not available
2023-11-28T11:14:31Z main - Command timeout set for 1 seconds
2023-11-28T11:14:31Z main - Using Legacy CLI to serve the command. (reason: unknown command "test" for "snyk")
2023-11-28T11:14:31Z legacycli:1 - Arguments: [test --all-projects -d]
2023-11-28T11:14:31Z legacycli:1 - Use StdIO: true
2023-11-28T11:14:31Z legacycli:1 - Working directory:
2023-11-28T11:14:31Z legacycli:1 - Init start
2023-11-28T11:14:31Z legacycli:1 - Init-Lock acquired: true (/Users/bdoetsch/Library/Caches/snyk/snyk-cli/1.1256.0-dev.9067b9ee7f7407151c23f2bde2742ecc1f0c4440.lock)
2023-11-28T11:14:31Z legacycli:1 - Validating sha256 of /Users/bdoetsch/Library/Caches/snyk/snyk-cli/1.1256.0-dev.9067b9ee7f7407151c23f2bde2742ecc1f0c4440/snyk-macos
2023-11-28T11:14:31Z legacycli:1 - open /Users/bdoetsch/Library/Caches/snyk/snyk-cli/1.1256.0-dev.9067b9ee7f7407151c23f2bde2742ecc1f0c4440/snyk-macos: no such file or directory
2023-11-28T11:14:31Z legacycli:1 - Extract cliv1 to /Users/bdoetsch/Library/Caches/snyk/snyk-cli/1.1256.0-dev.9067b9ee7f7407151c23f2bde2742ecc1f0c4440/snyk-macos
2023-11-28T11:14:31Z legacycli:1 - Validating sha256 of /Users/bdoetsch/Library/Caches/snyk/snyk-cli/1.1256.0-dev.9067b9ee7f7407151c23f2bde2742ecc1f0c4440/snyk-macos
2023-11-28T11:14:31Z legacycli:1 - expected:  14a44145b1fe71f5609ae10582bcb360f9c01ef0938ee4a8593fd1d7c5b8eac1
2023-11-28T11:14:31Z legacycli:1 - actual:    14a44145b1fe71f5609ae10582bcb360f9c01ef0938ee4a8593fd1d7c5b8eac1
2023-11-28T11:14:31Z legacycli:1 - Extracted cliv1 successfully
2023-11-28T11:14:31Z legacycli:1 - Init end
2023-11-28T11:14:31Z legacycli:1 - Temporary CertificateLocation: /Users/bdoetsch/Library/Caches/snyk/snyk-cli/1.1256.0-dev.9067b9ee7f7407151c23f2bde2742ecc1f0c4440/tmp/snyk-cli-cert-1900104616.crt
2023-11-28T11:14:31Z legacycli:1 - Enabled Proxy Authentication Mechanism: Anyauth
2023-11-28T11:14:31Z legacycli:1 - starting proxy
2023-11-28T11:14:31Z legacycli:1 - Wrapper proxy is listening on port:  60154
2023-11-28T11:14:31Z legacycli:1 - Launching:
2023-11-28T11:14:31Z legacycli:1 - /Users/bdoetsch/Library/Caches/snyk/snyk-cli/1.1256.0-dev.9067b9ee7f7407151c23f2bde2742ecc1f0c4440/snyk-macos
2023-11-28T11:14:31Z legacycli:1 - With Arguments:
2023-11-28T11:14:31Z legacycli:1 - test, --all-projects, -d
2023-11-28T11:14:31Z legacycli:1 - With Environment:
2023-11-28T11:14:31Z legacycli:1 - NODE_EXTRA_CA_CERTS = /Users/bdoetsch/Library/Caches/snyk/snyk-cli/1.1256.0-dev.9067b9ee7f7407151c23f2bde2742ecc1f0c4440/tmp/snyk-cli-cert-1900104616.crt
2023-11-28T11:14:31Z legacycli:1 - HTTPS_PROXY = http://snykcli:e5d19da8-84ab-4ae4-a756-1ab706dce114@127.0.0.1:60154
2023-11-28T11:14:31Z legacycli:1 - HTTP_PROXY = http://snykcli:e5d19da8-84ab-4ae4-a756-1ab706dce114@127.0.0.1:60154
2023-11-28T11:14:31Z legacycli:1 - NO_PROXY = localhost,127.0.0.1,::1
2023-11-28T11:14:31Z legacycli:1 - SNYK_SYSTEM_HTTPS_PROXY =
2023-11-28T11:14:31Z legacycli:1 - SNYK_SYSTEM_HTTP_PROXY =
2023-11-28T11:14:31Z legacycli:1 - SNYK_SYSTEM_NO_PROXY =
command timed out
2023-11-28T11:14:32Z legacycli:1 - Proxy successfully shut down
2023-11-28T11:14:32Z legacycli:1 - deleting temp cert file: /Users/bdoetsch/Library/Caches/snyk/snyk-cli/1.1256.0-dev.9067b9ee7f7407151c23f2bde2742ecc1f0c4440/tmp/snyk-cli-cert-1900104616.crt
2023-11-28T11:14:32Z legacycli:1 - deleted temp cert file: /Users/bdoetsch/Library/Caches/snyk/snyk-cli/1.1256.0-dev.9067b9ee7f7407151c23f2bde2742ecc1f0c4440/tmp/snyk-cli-cert-1900104616.crt
2023-11-28T11:14:32Z main - Exiting with 69
2023-11-28T11:14:32Z main - Sending Analytics
2023-11-28T11:14:33Z main - Analytics successfully send
```

## Where should the reviewer start?
`cliv2/internal/cliv2/cliv2.go`
`cliv2/cmd/cliv2/main.go`

## How should this be manually tested?
- Build with `make build`
- `SNYK_CI=1 SNYK_TIMEOUT_SECS=1 binary-releases/snyk-macos test --all-projects` should exit with exitCode 69 and "command timed out" message

## Automatic tests
- https://github.com/snyk/cli/blob/9067b9ee7f7407151c23f2bde2742ecc1f0c4440/cliv2/cmd/cliv2/main_test.go#L217
- https://github.com/snyk/cli/blob/9067b9ee7f7407151c23f2bde2742ecc1f0c4440/cliv2/internal/cliv2/cliv2_test.go#L467

### CODE REVIEW

File: cliv2/internal/cliv2/cliv2_test.go [hunk 1]
```diff
@@ -460,3 +463,22 @@ func Test_clearCacheBigCache(t *testing.T) {
...
 	assert.FileExists(t, currentVersion)
 }
+
+func Test_setTimeout(t *testing.T) {
+	if //goland:noinspection ALL
+	runtime.GOOS == "windows" {
+		t.Skip("Skipping test on windows")
+	}
+	config := configuration.NewInMemory()
+	cli, _ := cliv2.NewCLIv2(config, discardLogger)
```
Comment (cmars) [inline cliv2/internal/cliv2/cliv2_test.go:473]:
```suggestion
	cli, err := cliv2.NewCLIv2(config, discardLogger)
	assert.NoError(t, err)
```

File: cliv2/internal/cliv2/cliv2_test.go [hunk 2]
```diff
@@ -460,3 +463,22 @@ func Test_clearCacheBigCache(t *testing.T) {
 	assert.DirExists(t, dir6)
 	assert.FileExists(t, currentVersion)
 }
+
+func Test_setTimeout(t *testing.T) {
+	if //goland:noinspection ALL
```
Comment (cmars) [inline cliv2/internal/cliv2/cliv2_test.go:468]: Do we need IDE-specific comments like this? Does this test pose a problem for Windows because of the SIGKILL? Could we use `os/Process.Kill` as an alternative on Windows?

File: cliv2/internal/cliv2/cliv2_test.go [hunk 3]
```diff
@@ -310,7 +313,7 @@ func Test_extractOnlyOnce(t *testing.T) {
 
 	// run twice
 	assert.Nil(t, cli.Init())
-	cli.Execute(getProxyInfoForTest(), []string{"--help"})
+	_ = cli.Execute(getProxyInfoForTest(), []string{"--help"})
```
Comment (cmars) [inline cliv2/internal/cliv2/cliv2_test.go:316]: Error can be asserted nil.

File: cliv2/internal/cliv2/cliv2_test.go [hunk 4]
```diff
@@ -301,7 +304,7 @@ func Test_extractOnlyOnce(t *testing.T) {
 
 	// run once
 	assert.Nil(t, cli.Init())
-	cli.Execute(getProxyInfoForTest(), []string{"--help"})
+	_ = cli.Execute(getProxyInfoForTest(), []string{"--help"})
```
Comment (cmars) [inline cliv2/internal/cliv2/cliv2_test.go:307]: Error can be asserted nil here.

File: cliv2/cmd/cliv2/main_test.go [hunk 5]
```diff
@@ -1,16 +1,15 @@
...
-
 	"github.com/snyk/go-application-framework/pkg/configuration"
 	localworkflows "github.com/snyk/go-application-framework/pkg/local_workflows"
 	"github.com/snyk/go-application-framework/pkg/workflow"
 	"github.com/spf13/cobra"
 	"github.com/spf13/pflag"
 	"github.com/stretchr/testify/assert"
+	"os"
+	"testing"
+	"time"
```
Comment (cmars) [inline cliv2/cmd/cliv2/main_test.go:12]: Grouping the imports into sections is common practice; first stdlib, then 3rd party, then project-local.

There is usually a golangci-lint rule that checks this (gci), but I'm not sure whether it is set up for this project.

File: cliv2/cmd/cliv2/main.go [hunk 6]
```diff
@@ -479,8 +486,9 @@ func setTimeout(config configuration.Configuration, onTimeout func()) {
 	}
 	debugLogger.Printf("Command timeout set for %d seconds", timeout)
 	go func() {
-		<-time.After(time.Duration(timeout) * time.Second)
-		fmt.Fprintf(os.Stderr, "command timed out\n")
+		// we wait a bit longer than the timeout to ensure that the command has enough time to finish
+		<-time.After(time.Duration(timeout+5) * time.Second)
```
Comment (cmars) [inline cliv2/cmd/cliv2/main.go:490]: Could we lower this to 3 seconds? Or even 1 second?

Could we use a constant for this extra delay?

File: cliv2/internal/cliv2/cliv2.go [hunk 7]
```diff
@@ -235,8 +243,8 @@ func (c *CLI) commandAbout(proxyInfo *proxy.ProxyInfo, passthroughArgs []string)
 			}
 
 			fmt.Printf("Package: %s \n", strings.ReplaceAll(strings.ReplaceAll(fPath, "/licenses/", ""), "/"+f.Name(), ""))
-			fmt.Fprintln(c.stdout, string(data))
-			fmt.Fprint(c.stdout, separator)
+			_, _ = fmt.Fprintln(c.stdout, string(data))
+			_, _ = fmt.Fprint(c.stdout, separator)
```
Comment (cmars) [inline cliv2/internal/cliv2/cliv2.go:247]: Wondering why this change is made here? Is it because a golangci-lint rule is complaining about it?

If we're not doing anything with the bytes written or the error I'd be inclined to write a utility func that does an unconditional stderr print. Then discard the unused return values once, in that func.

File: cliv2/internal/cliv2/cliv2.go [hunk 8]
```diff
@@ -340,16 +348,9 @@ func PrepareV1EnvironmentVariables(
...
 }
 
-func (c *CLI) PrepareV1Command(
-	cmd string,
-	args []string,
-	proxyInfo *proxy.ProxyInfo,
-	integrationName string,
-	integrationVersion string,
-) (snykCmd *exec.Cmd, err error) {
+func (c *CLI) PrepareV1Command(ctx context.Context, cmd string, args []string, proxyInfo *proxy.ProxyInfo, integrationName string, integrationVersion string) (snykCmd *exec.Cmd, err error) {
```
Comment (cmars) [inline cliv2/internal/cliv2/cliv2.go:351]: The prior format was easier to read IMO.

File: cliv2/internal/cliv2/cliv2.go [hunk 9]
```diff
@@ -360,8 +361,18 @@ func (c *CLI) PrepareV1Command(
...
 
 func (c *CLI) executeV1Default(proxyInfo *proxy.ProxyInfo, passThroughArgs []string) error {
+	timeout := c.globalConfig.GetInt(configuration.TIMEOUT)
+	var ctx context.Context
+	var cancel context.CancelFunc
+	if timeout == 0 {
+		ctx = context.Background()
+	} else {
+		deadline := time.Now().Add(time.Duration(timeout) * time.Second)
+		ctx, cancel = context.WithDeadline(context.Background(), deadline)
```
Comment (cmars) [inline cliv2/internal/cliv2/cliv2.go:371]: Suggesting a slightly more concise form,

```suggestion
		ctx, cancel = context.WithTimeout(context.Background(), time.Duration(timeout) * time.Second)
```

File: cliv2/internal/cliv2/cliv2.go [hunk 10]
```diff
@@ -397,13 +408,16 @@ func (c *CLI) executeV1Default(proxyInfo *proxy.ProxyInfo, passThroughArgs []str
...
+		if errors.As(err, &evWarning) {
+			_, _ = fmt.Fprintln(c.stdout, "WARNING! ", evWarning)
 		}
 	}
 
 	err = snykCmd.Run()
-
+	if errors.Is(ctx.Err(), context.DeadlineExceeded) {
+		return ctx.Err()
+	}
```
Comment (cmars) [inline cliv2/internal/cliv2/cliv2.go:420]: Wondering if the `err` returned by Run above would satisfy this; having to check the context error explicitly like this is unusual (but could be necessary depending on how Run is implemented...).

**Review by cmars — COMMENTED** [review]
Using a context is a huge improvement here and this is a great catch!

Strongly suggest that `CLI.PrepareV1Command` use [exec.CommandContext](https://pkg.go.dev/os/exec#CommandContext) -- then the context timeout will automatically tear down the node subprocess. 

If you do this, it should not be necessary to explicitly check the `ctx.Err` after running node; this will be handled for you by the `os/exec.Cmd`, see https://cs.opensource.google/go/go/+/refs/tags/go1.21.4:src/os/exec/exec.go;l=665

Aside from that, I would strongly recommend checking errors in tests. Discarding them for unconditional stderr printing is OK.

File: cliv2/internal/cliv2/cliv2_test.go (hunk 2 above)
Comment (cmars) [reply cliv2/internal/cliv2/cliv2_test.go:468]: Followup on this... I think `exec.CommandContext` may cover the need for this (but it should be tested)

File: cliv2/cmd/cliv2/main.go (hunk 6 above)
Comment (bastiandoetsch) [reply cliv2/cmd/cliv2/main.go:490]: Sure. Lower than 3s did not work for me reliably.

File: cliv2/cmd/cliv2/main_test.go (hunk 5 above)
Comment (bastiandoetsch) [reply cliv2/cmd/cliv2/main_test.go:12]: yeah, golangci-lint is not used in this project

File: cliv2/internal/cliv2/cliv2.go (hunk 9 above)
Comment (bastiandoetsch) [reply cliv2/internal/cliv2/cliv2.go:371]: Why `withTimeout` instead of `withDeadline`?

File: cliv2/internal/cliv2/cliv2_test.go (hunk 4 above)
Comment (bastiandoetsch) [reply cliv2/internal/cliv2/cliv2_test.go:307]: That may be a change in the expected behaviour in the test. It was implicitly ignoring it before, now it is explicitly ignoring it.

Comment (cmars) [reply cliv2/internal/cliv2/cliv2_test.go:307]: That's fair.. realizing now that `--help` may exit non-zero.

I'd like to tidy this up a bit in a followup to assert expected behavior, but let's not block on it. Sorry for the noise.

**Comment by bastiandoetsch** [comment]
> Using a context is a huge improvement here and this is a great catch!
> 
> Strongly suggest that `CLI.PrepareV1Command` use [exec.CommandContext](https://pkg.go.dev/os/exec#CommandContext) -- then the context timeout will automatically tear down the node subprocess.
> 
> If you do this, it should not be necessary to explicitly check the `ctx.Err` after running node; this will be handled for you by the `os/exec.Cmd`, see https://cs.opensource.google/go/go/+/refs/tags/go1.21.4:src/os/exec/exec.go;l=665
> 
> Aside from that, I would strongly recommend checking errors in tests. Discarding them for unconditional stderr printing is OK.

Hm. I don't understand what's the difference to the current implementation. PrepareV1Command is already using CommandContext, so what is the suggested change?

File: cliv2/internal/cliv2/cliv2_test.go (hunk 2 above)
Comment (cmars) [reply cliv2/internal/cliv2/cliv2_test.go:468]: Thanks for the context!

Comment (cmars) [reply cliv2/internal/cliv2/cliv2_test.go:468]: (Pun not intended but totally should have been :D)

**Review by cmars — APPROVED** [review]
LGTM

**Comment by cmars** [comment]
> Hm. I don't understand what's the difference to the current implementation. PrepareV1Command is already using CommandContext, so what is the suggested change?

I missed that change in this branch, sorry!

**Review by dotkas — APPROVED** [review]
LGTM, agreeing with Casey's comment about https://github.com/snyk/cli/pull/4950/files#r1407783863 though. I usually set Golang to ignore that specific error as I don't agree with it 🤣

File: cliv2/internal/cliv2/cliv2.go (hunk 10 above)
Comment (bastiandoetsch) [reply cliv2/internal/cliv2/cliv2.go:426]: I agree it is weird. Maybe something to look into later.
